README.md
Flutter SVG Icons

`
A package to show svg asset files in icon widgets, supporting responsive svg icons.
`

## Features

* Show svg asset files in icon widgets.
* Support responsive icons.


## Getting started
In pubspec.yaml file, insert `flutter_svg_icons` as dependency:
```
  flutter_svg_icons: ^0.0.1
```



## Usage

Make sure that your svg files are added to your project assets.

* Svg Icon - by default, its color will be the same as icon theme color.
```dart
SvgIcon(icon: SvgIconData('assets/flutter_logo.svg'))
```

* Svg Icon - disable `responsiveColor` and set custom `color`.
```dart
SvgIcon(color: Colors.red, responsiveColor:false, icon: SvgIconData('assets/flutter_logo.svg'))
```

* Responsive Svg Icon
```dart
SvgIcon(icon: ResponsiceSvgIconData({
      64: 'assets/archive_s.svg',
      128: 'assets/archive_m.svg',
      9999: 'assets/archive_l.svg'
    }))
```

CHANGELOG.md
## 0.0.1
* First release of `SVG Icons` package.


lib/src/svg_icon.dart
part of flutter_svg_icons;

/// An icon widget shows asset svg file.
class SvgIcon extends StatelessWidget {
  /// The svg icon data of this widget
  final SvgIconData icon;

  /// Weather the color of the icon following the icon theme color
  final bool responsiveColor;

  /// The color of this icon, [responsiveColor] must be false
  final Color? color;

  /// The size of this icon
  final double? size;

  /// Create an instance of [SvgIcon]
  const SvgIcon(
      {Key? key,
      required this.icon,
      this.responsiveColor = true,
      this.size,
      this.color})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    if (icon is ResponsiveSvgIconData) {
      {
        Widget result = _ResponsiveSvgIcon(icon as ResponsiveSvgIconData,
            responsiveColor: responsiveColor, color: this.color);
        var size = this.size ?? IconTheme.of(context).size;
        if (size != null) {
          result = SizedBox(
            width: size,
            height: size,
            child: result,
          );
        }
        return result;
      }
    }
    var color = !responsiveColor ? this.color : IconTheme.of(context).color;
    Widget result = SvgPicture.asset(
      icon.fullPath,
      package: icon.package,
      color: color,
      width: size,
      height: size,
    );

    bool mirrored = false;

    if (icon.reflectable == true) {
      if (icon.reflected != true) {
        mirrored = (Directionality.of(context) == TextDirection.rtl);
      } else {
        mirrored = (Directionality.of(context) == TextDirection.ltr);
      }
    } else {
      mirrored = (icon.reflected == true);
    }

    if (mirrored) {
      result = Transform(
        alignment: Alignment.center,
        transform: Matrix4.rotationY(pi),
        child: result,
      );
    }

    return result;
  }
}


lib/src/svg_icon_data.dart
part of flutter_svg_icons;

/// Represents the svg icon data to be shown in [SvgIcon] widget
class SvgIconData {
  /// The asset package
  final String? package;

  /// The asset prefix
  final String? prefix;

  /// The asset file name, (without .svg)
  final String name;

  /// Weather the icon is following the build context text direction
  final bool? reflectable;

  /// Weather to reflect this svg icon
  final bool? reflected;

  /// Create an instance of [SvgIconData]
  const SvgIconData(this.name,
      {this.package, this.prefix, this.reflectable, this.reflected});

  /// Get the full path of this icon svg file
  String get fullPath => prefix == null ? name : '$prefix/$name';

  /// Create another instance of [SvgIconData] with specified configurations.
  SvgIconData copyWith({bool? reflectable, bool? reflected}) =>
      SvgIconData(name,
          prefix: prefix,
          reflectable: reflectable ?? this.reflectable,
          reflected: reflected ?? this.reflected);
}


lib/src/responsive_svg_icon.dart
// ignore_for_file: non_constant_identifier_names, constant_identifier_names
part of flutter_svg_icons;

class _ResponsiveSvgIcon extends StatelessWidget {
  final ResponsiveSvgIconData icon;
  final bool responsiveColor;
  final Color? color;

  const _ResponsiveSvgIcon(this.icon,
      {Key? key, this.responsiveColor = true, this.color})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return LayoutBuilder(builder: (c, cs) {
      var dim = min(cs.maxWidth, cs.maxHeight);

      String? path;

      for (var d in icon.assetPaths.keys) {
        if (dim <= d) {
          path = icon.assetPaths[d];
          break;
        }
      }

      return path == null
          ? const SizedBox()
          : SvgIcon(
              icon: SvgIconData(path,
                  reflectable: icon.reflectable, reflected: icon.reflected),
              size: dim,
              color: color,
              responsiveColor: responsiveColor,
            );
    });
  }
}


lib/src/responsive_svg_icon_data.dart
part of flutter_svg_icons;

/// Represents a svg icon data, with several asset files to shown
/// according to the allowed size.
class ResponsiveSvgIconData extends SvgIconData {
  /// Represesnts the multiple variations.
  /// ```dart
  /// assetPaths = {
  ///  64: 'assets/archive_s.svg',
  ///  128: 'assets/archive_m.svg',
  ///  9999: 'assets/archive_l.svg'
  /// }
  /// ```
  final Map<double, String> assetPaths;

  /// Create an instance of [ResponsiveSvgIconData]
  const ResponsiveSvgIconData(this.assetPaths,
      {String? package, bool? reflectable, bool? reflected})
      : super('',
            package: package, reflectable: reflectable, reflected: reflected);

  @override

  /// Create another instance of [ResponsiveSvgIconData] with specified configurations.
  ResponsiveSvgIconData copyWith({bool? reflectable, bool? reflected}) =>
      ResponsiveSvgIconData(assetPaths,
          package: package,
          reflectable: reflectable ?? this.reflectable,
          reflected: reflected ?? this.reflected);
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_svg_icons/flutter_svg_icons.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'SVG Icons',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'SVG Icons'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    var responsiveIconData = ResponsiveSvgIconData({
      64: 'assets/archive_s.svg',
      128: 'assets/archive_m.svg',
      9999: 'assets/archive_l.svg'
    });
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        children: [
          const Text(
            'As icon theme color:',
            style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
          ),
          const SvgIcon(size: 60, icon: SvgIconData('assets/flutter_logo.svg')),
          const Text(
            'Pure color:',
            style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
          ),
          const SvgIcon(
              size: 60,
              responsiveColor: false,
              icon: SvgIconData(
                'assets/flutter_logo.svg',
              )),
          const Text(
            'Customized color:',
            style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
          ),
          const SvgIcon(
              responsiveColor: false,
              size: 60,
              color: Colors.red,
              icon: SvgIconData(
                'assets/flutter_logo.svg',
              )),
          const Text(
            'Responsive Icon:',
            style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
          ),
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,
            children: [
              SvgIcon(
                icon: responsiveIconData,
                size: 25,
              ),
              SvgIcon(
                icon: responsiveIconData,
                size: 70,
              ),
              SvgIcon(
                icon: responsiveIconData,
                size: 130,
              )
            ],
          )
        ],
      ),
    );
  }
}
